safenode/src/protocol/types/transfers/mod.rs

```diff
@@ -0,0 +1,225 @@
+    let dbc_builder = tx_builder
+        .build(Hash::default(), &mut rng)
+        .map_err(Error::Dbcs)?;
+
+    // Perform verifications of input TX and spentproofs,
+    // as well as building the output DBCs.
+    let mut created_dbcs: Vec<_> = dbc_builder
+        .build()
+        .map_err(Error::Dbcs)?
+        .into_iter()
+        .map(|(dbc, amount)| CreatedDbc { dbc, amount })
+        .collect();
+
+    let mut change_dbc = None;
+    created_dbcs.retain(|created| {
+        if created.dbc.public_address() == &change_to && change.as_nano() > 0 {
+            change_dbc = Some(created.dbc.clone());
+            false
+        } else {
+            true
+        }
+    });
+
+    Ok(Outputs {
+        created_dbcs,
+        change_dbc,
+    })
+}
+
+/// Select the necessary number of dbcs from those that we were passed.
+#[allow(clippy::result_large_err)]
+pub(crate) fn select_inputs(
+    inputs: Vec<(Dbc, DerivedKey)>,
+    recipients: Vec<(Token, DbcIdSource)>,
+    change_to: PublicAddress,
+) -> Result<Inputs> {
+    let mut dbcs_to_spend = Vec::new();
+    let mut total_input_amount = Token::zero();
+    let total_output_amount = recipients
+        .iter()
+        .fold(Some(Token::zero()), |total, (amount, _)| {
+            total.and_then(|t| t.checked_add(*amount))
+        })
+        .ok_or_else(|| {
+            Error::DbcReissueFailed(
+                "Overflow occurred while summing the amounts for the recipients.".to_string(),
+            )
+        })?;
+
+    let mut change_amount = total_output_amount;
+
+    for (dbc, derived_key) in inputs {
+        let input_key = dbc.id();
+
+        let dbc_balance = match dbc.revealed_amount(&derived_key) {
+            Ok(revealed_amount) => Token::from_nano(revealed_amount.value()),
+            Err(err) => {
+                warn!("Ignoring input Dbc (id: {input_key:?}) due to not having correct derived key: {err:?}");
+                continue;
+            }
+        };
+
+        // Add this Dbc as input to be spent.
+        dbcs_to_spend.push((dbc, derived_key));
+
+        // Input amount increases with the amount of the dbc.
+        total_input_amount = total_input_amount.checked_add(dbc_balance)
+            .ok_or_else(|| {
+                Error::DbcReissueFailed(
+                    "Overflow occurred while increasing total input amount while trying to cover the output DBCs."
+                    .to_string(),
+            )
+            })?;
+
+        // If we've already combined input DBCs for the total output amount, then stop.
+        match change_amount.checked_sub(dbc_balance) {
+            Some(pending_output) => {
+                change_amount = pending_output;
+                if change_amount.as_nano() == 0 {
+                    break;
+                }
+            }
+            None => {
+                change_amount = Token::from_nano(dbc_balance.as_nano() - change_amount.as_nano());
```

Comment:
bit confused here, when an overflow happens, this means the `dbc_balance` is greater than our left un-paid change,
which, the `change_aount` can be fulfill (i.e. set to 0 ?)

Reply:
Yeah, I'm actually not entirely sure about this line myself. It is a copy paste from an older implementation.